From a0e77361d51fdac58e321a873d4a82377bf28968 Mon Sep 17 00:00:00 2001
From: Tim-cy Yang <Tim-cy.Yang@mediatek.com>
Date: Mon, 15 Sep 2025 10:27:09 +0800
Subject: [PATCH 12/23] board: mediatek: common: add verification check before
 upgrading

[Description]
add verification check vefore upgrading. Check the signatures for all
configuration nodes using the public key placed in u-boot dtb when
secure boot is enabled.

[Release-Log]
N/A

Change-Id: I046ffe49e986f0b2504f23984d6e7ac75e38e8c0
Signed-off-by: Tim-cy Yang <Tim-cy.Yang@mediatek.com>
CR-Id: WCNCR00402800
---
 board/mediatek/Kconfig                |  1 +
 board/mediatek/common/verify_helper.c | 55 +++++++++++++++++++++++++++
 2 files changed, 56 insertions(+)

diff --git a/board/mediatek/Kconfig b/board/mediatek/Kconfig
index f0b28dc07ff..7c70acbfbc9 100644
--- a/board/mediatek/Kconfig
+++ b/board/mediatek/Kconfig
@@ -304,6 +304,7 @@ menuconfig MTK_SECURE_BOOT
 	select CHAIN_OF_TRUST if !MTK_DUAL_BOOT
 	select IMAGE_FORCED_VERIFY
 	select FDT_NO_BOOTARGS_OVERRIDE
+	select MTK_UPGRADE_IMAGE_VERIFY
 	depends on ARCH_MEDIATEK
 	default n
 	help
diff --git a/board/mediatek/common/verify_helper.c b/board/mediatek/common/verify_helper.c
index 7b510522fbd..596bd7794ae 100644
--- a/board/mediatek/common/verify_helper.c
+++ b/board/mediatek/common/verify_helper.c
@@ -105,6 +105,47 @@ static bool fit_record_hashes(const void *fit, int image_noffset,
 	return hashes->flags != 0;
 }
 
+#ifdef CONFIG_MTK_SECURE_BOOT
+static int fit_all_config_verify(const void *fit)
+{
+	int confs_noffset;
+	int noffset;
+	int ndepth;
+	int count;
+
+	/* Find configurations parent node offset */
+	confs_noffset = fdt_path_offset(fit, FIT_CONFS_PATH);
+	if (confs_noffset < 0) {
+		printf("Can't get configurations parent node '%s' (%s)\n",
+		       FIT_CONFS_PATH, fdt_strerror(confs_noffset));
+		return 0;
+	}
+
+	/* Process its subnodes, print out configurations details */
+	printf("## Checking hash(es) integrity for FIT Image at %08lx ...\n",
+	       (ulong)fit);
+	for (ndepth = 0, count = 0,
+		noffset = fdt_next_node(fit, confs_noffset, &ndepth);
+			(noffset >= 0) && (ndepth > 0);
+			noffset = fdt_next_node(fit, noffset, &ndepth)) {
+		if (ndepth == 1) {
+			/*
+			 * Direct child node of the configurations parent node,
+			 * i.e. configuration node.
+			 */
+			printf("   Hash(es) Integrity for Configuration %u (%s): ",
+			       count++, fit_get_name(fit, noffset, NULL));
+
+			if (fit_config_verify(fit, noffset))
+				return 0;
+			printf("\n");
+		}
+	}
+
+	return 1;
+}
+#endif /* CONFIG_MTK_SECURE_BOOT */
+
 /**
  * verify_kernel_fit() - Verify kernel FIT image
  *
@@ -126,6 +167,13 @@ static bool verify_kernel_fit(const void *fit,  size_t size,
 		return false;
 	}
 
+#ifdef CONFIG_MTK_SECURE_BOOT
+	if (!fit_all_config_verify(fit)) {
+		printf("FIT image configuration integrity checking failed\n");
+		return false;
+	}
+#endif
+
 	if (!fit_all_image_verify(fit)) {
 		printf("FIT image integrity checking failed\n");
 		return false;
@@ -435,6 +483,13 @@ bool verify_image_ram(const void *data, size_t size, u32 block_size,
 			return false;
 		}
 
+#ifdef CONFIG_MTK_SECURE_BOOT
+		if (!fit_all_config_verify(data)) {
+			printf("FIT image configuration integrity checking failed\n");
+			return false;
+		}
+#endif
+
 		if (!fit_all_image_verify(data)) {
 			printf("FIT image integrity checking failed\n");
 			return false;
-- 
2.45.2

